### DISCUSSION TITLE
Pthread: not working on webworker.js

### DISCUSSION THREAD

**Discussion by antony0604** [emscripten-core/emscripten#17561] [Q&A]
I had read the follow reference about running pthread on webwork.js :
https://groups.google.com/g/emscripten-discuss/c/EO1q0lDKi6Y/m/LyB2Hro1CgAJ
https://github.com/emscripten-core/emscripten/issues/11914
https://github.com/emscripten-core/emscripten/pull/12102

My emsdk is newest for now (3.1.18), so it should had fixed the issue 12102.
but i still cann't get pthread be create on main runtime thread. ( it can be create on main browser thread)

// follow are the source code of html/js

//test_script.html
```
<html>
<head>
</head>
<body>
    <script>
    /*
    Module = {};
    Module.onRuntimeInitialized = function() {
      console.log("Module onRuntimeInitialized...");
    }
    */
    </script>
    <script src="script.js"></script>
</body>
```

// script.js
`const worker = new Worker("worker.js");`

// worker.js
```
self.Module = {
    onRuntimeInitialized: function () {
        onWasmLoaded();
    }
};
self.importScripts("multi_thread.js");
function onWasmLoaded() {
        console.log("onWasmLoaded ...");
}
```

//multi_thread.c
```
//emcc multi_thread.c  -O3 -s USE_PTHREADS=1  -s PROXY_TO_PTHREAD -s EXIT_RUNTIME=1 -o multi_thread.js
#include <stdio.h>
#include <pthread.h>

#define NUMTHRDS 2

typedef struct arg_struct {
    int id;
    int val;
} arg_struct;

void *printarg(void *params) {
    arg_struct *args = (arg_struct *) params;
    printf("id %i value %i\n", args->id, args->val);
    return 0;
}

int createThread() {
    printf("createThread begin...\n");
    int i = 0;
    pthread_t threads[NUMTHRDS];
    arg_struct args[NUMTHRDS];
    for (i = 0; i < NUMTHRDS; i++) {
        args[i].id = i;
        args[i].val = i + 10;
        pthread_create(&(threads[i]), NULL, &printarg, &args[i]);
    }
    for (i = 0; i < NUMTHRDS; i++) {
        pthread_join(threads[i], NULL);
    }
    printf("createThread end...\n");
	return 0;
}

int main() {
    printf("main start...\n");
    createThread();
    printf("main end...\n");
    return 0;
}

```
main_browser.png show on html environment, to call pthread_create, it work fine
![main_browser](https://user-images.githubusercontent.com/110456487/182340164-07d10712-9f02-4983-bff2-92e579be35c9.png)

but main_runtime_thread.png show on web worker environment, to call pthread_create, it will failed.
![main_runtime_thread](https://user-images.githubusercontent.com/110456487/182340152-07b37099-16dd-4705-b995-7517f0020105.png)

Could anyone help me to finger out what cause this issue?

**Comment by sbc100** [comment]
Are you sure you are not running into https://emscripten.org/docs/porting/pthreads.html#blocking-on-the-main-browser-thread?

You need to return to event loop before you threads will start... or you can use PTHREAD_POOL_SIZE as a workaround.

**Reply by sbc100** [reply]
Sorry I meant https://emscripten.org/docs/porting/pthreads.html#special-considerations

**Comment by antony0604** [comment]
The compile command is as follow:
emcc multi_thread.c  -O3 -s USE_PTHREADS=1  -s PROXY_TO_PTHREAD -s EXIT_RUNTIME=1 -o multi_thread.js
The command contain PROXY_TO_PTHREAD , so it can create and execute thread if multi_thread.js is run on html file.

The multi_thread.c is the same as the first post of this issue.

The browser infomation: Chrome  104.0.5112.81 64bit , run with (--enable-features=SharedArrayBuffer)

If run multi_thread,js on main brower thread (test_script.html ) , it will work fine to create thread

// test_script.html
```
<html>
<head>
</head>
<body>
    <script>   
    Module = {};
    Module.onRuntimeInitialized = function() {
      console.log("Module onRuntimeInitialized...");
    }   
    </script>
    <script src="script.js"></script>
    <script src="multi_thread.js"></script>
</body>

```
// script.js 
`const worker = new Worker("worker.js")`

// work.js
```
function onWasmLoaded() {
        console.log("onWasmLoaded ...");
}

```


But if start multi_thread.js on a webwork(work.js), it doesn't work. it can't create and execute the thread.

// test_scripte.html
```
<html>
<head>
</head>
<body>
    <script>
    /*
    Module = {};
    Module.onRuntimeInitialized = function() {
      console.log("Module onRuntimeInitialized...");
    }
    */
    </script>
    <script src="script.js"></script>
</body>
```

// script.js
`const worker = new Worker("worker.js");`

// work.js
```
// ptherad does not work in worker.js

self.Module = {
    onRuntimeInitialized: function () {
        onWasmLoaded();
    }
};

self.importScripts("multi_thread.js");

function onWasmLoaded() {
        console.log("onWasmLoaded ...");
}

```
